homeassistant/components/light/tcpbulbs.py

```diff
@@ -0,0 +1,146 @@
+"""
+Support for TCP Connected (Greenwave Reality) lights.
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/light.tcpbulbs/
+"""
+import logging
+import voluptuous as vol
+# Import the device class from the component that you want to support
+from homeassistant.components.light import (
+    ATTR_BRIGHTNESS, Light, PLATFORM_SCHEMA, SUPPORT_BRIGHTNESS)
+from homeassistant.const import CONF_HOST
+import homeassistant.helpers.config_validation as cv
+# Home Assistant depends on 3rd party packages for API specific code.
+SUPPORTED_FEATURES = (SUPPORT_BRIGHTNESS)
+_LOGGER = logging.getLogger(__name__)
+# Validation of the user's configuration
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+})
+
+
+def pullXML(host, bulbxml):
+    REQUIREMENTS = ['xmltodict']
+    import xmltodict
+    resp = tcpAction(host, 0, 'xml', 0)
+    # Store bulb information as xml file
+    with open(bulbxml, 'wb') as f:
+        f.write(resp.content)
+
+    with open(bulbxml) as fd:
+        doc = xmltodict.parse(fd.read())
+        return doc
+
+def numMap(value, leftMin, leftMax, rightMin, rightMax):
+    # Figure out how 'wide' each range is
+    leftSpan = leftMax - leftMin
+    rightSpan = rightMax - rightMin
+
+    # Convert the left range into a 0-1 range (float)
+    valueScaled = float(value - leftMin) / float(leftSpan)
+
+    # Convert the 0-1 range into a value in the right range.
+    return int(rightMin + (valueScaled * rightSpan))
+
+def getBrightness(device):
+        if 'level' in device:
+            level = numMap(int(device['level']),1,100,1,255)
+            return(level)
+        else:
+            return(0)
+            
+def checkOnline(device):
+        if 'offline' in device:
+            return(False)
+        else:
+            return(True)
+
+def tcpAction(host, did, action, value):
+    import requests
+    if action == 'xml':
+        url = 'http://' + host + '/gwr/gop.php?cmd=GWRBatch&data=<gwrcmds><gwrcmd><gcmd>RoomGetCarousel</gcmd><gdata><gip><version>1</version><token>1234567890</token><fields>name,status</fields></gip></gdata></gwrcmd></gwrcmds>&fmt=xml'

```

Comment:
I think the URLs (or at least the XML payloads) should be constructed separately. Currently this is looks quite wild.

Reply:
The XML URLs are quite wild, but theres nothing you can really do as that's the only way the gateway accepts it.

Reply:
You can use string concatenation to make it simpler. However, the question still remains whether this is so device specific that it should go to a separate library that you can then `import`.

Reply:
@amelchio This is very device specific. The gateway requires the requests to be formed this way, and I don't know of anything else that would. Should I move this to its own library? and if so, how would I go about getting that somewhere where HomeAssistant can grab it

Reply:
The HA rule is that the device specific protocol goes into a separate library that you publish on pypi.python.org. But in this case it would be something like a 20 line library so I don't know, maybe the rule can be bent.

Reply:
It works as-is, I can leave it in, create a library, or do whatever the team decides.